elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/ElasticsearchSourceStage.scala

```diff
@@ -0,0 +1,144 @@
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new ElasticsearchSourceLogic[T](indexName, typeName, query, client, settings, out, shape) {
+      override protected def convert(jsObj: JsObject): T = jsObj.convertTo[T]
+    }
+
+}
+
+sealed abstract class ElasticsearchSourceLogic[T](indexName: String,
+                                                  typeName: String,
+                                                  query: String,
+                                                  client: RestClient,
+                                                  settings: ElasticsearchSourceSettings,
+                                                  out: Outlet[OutgoingMessage[T]],
+                                                  shape: SourceShape[OutgoingMessage[T]])
+    extends GraphStageLogic(shape) {
+
+  private var scrollId: String = null
+  private val buffer = new util.ArrayDeque[OutgoingMessage[T]]()
+
+  protected def convert(jsObj: JsObject): T
+
+  def receiveMessages(): Unit =
+    try {
+      val res = if (scrollId == null) {
+        client.performRequest(
+          "POST",
+          s"$indexName/$typeName/_search",
+          Map("scroll" -> "5m", "sort" -> "_doc").asJava,
+          new StringEntity(s"""{"size": ${settings.bufferSize}, "query": ${query}}""")
+        )
+      } else {
+        client.performRequest(
+          "POST",
+          s"/_search/scroll",
+          Map[String, String]().asJava,
+          new StringEntity(Map("scroll" -> "5m", "scroll_id" -> scrollId).toJson.toString)
+        )
+      }
+      handleSuccess(res)
+
+    } catch {
+      case ex: Exception => handleFailure(ex)
+    }
+
+  def handleFailure(ex: Exception): Unit =
+    failStage(ex)
+
+  def handleSuccess(res: Response): Unit = {
+    val json = {
+      val out = new ByteArrayOutputStream()
+      try {
+        res.getEntity.writeTo(out)
+        new String(out.toByteArray, "UTF-8")
+      } finally {
+        out.close()
+      }
+    }
+
+    val jsObj = json.parseJson.asJsObject
+
+    jsObj.fields.get("error") match {
+      case None => {
+        val hits = jsObj.fields("hits").asJsObject.fields("hits").asInstanceOf[JsArray]
+        if (hits.elements.isEmpty && scrollId != null) {
+          //completeStage()
+        } else {
+          scrollId = jsObj.fields("_scroll_id").asInstanceOf[JsString].value
+          hits.elements.reverse.foreach { element =>
+            val doc = element.asJsObject
+            val id = doc.fields("_id").asInstanceOf[JsString].value
+            val source = doc.fields("_source").asJsObject
+            buffer.addFirst(OutgoingMessage(id, convert(source)))
+          }
+        }
+      }
+      case Some(error) => {
+        failStage(new IllegalStateException(error.toString))
+      }
+    }
+  }
+
+  setHandler(out,
+    new OutHandler {
+    override def onPull(): Unit = {
+      if (buffer.isEmpty) {
```

Comment:
This can be simplified using `emit` to push the complete list of elements received by `receiveMessages`.